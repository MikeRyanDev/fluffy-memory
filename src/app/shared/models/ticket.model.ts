import { UserModel } from './user.model';

export interface TicketModel {
  id: number;
  description: string;
  assigneeId: number;
  completed: boolean;
}

export interface TicketViewModel extends TicketModel {
  assignee: UserModel;
}
