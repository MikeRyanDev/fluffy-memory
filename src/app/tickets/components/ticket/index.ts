export * from './ticket.component';
export * from './ticket.module';
