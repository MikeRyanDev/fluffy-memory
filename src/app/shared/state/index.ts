import { ActionReducerMap, createSelector } from '@ngrx/store';
import { CustomRouterState } from '../../custom-route-serializer';
import * as fromTicketDetailsPage from './ticket-details-page.reducer';
import * as fromTicketsPage from './tickets-page.reducer';
import * as fromTickets from './tickets.reducer';
import * as fromUsers from './users.reducer';
import { TicketViewModel } from '../models';
import { routerReducer, RouterReducerState } from '@ngrx/router-store';

export interface State {
  router: RouterReducerState<CustomRouterState>;
  ticketDetailsPage: fromTicketDetailsPage.State;
  ticketsPage: fromTicketsPage.State;
  tickets: fromTickets.State;
  users: fromUsers.State;
}

export const reducers: ActionReducerMap<State, any> = {
  router: routerReducer,
  ticketDetailsPage: fromTicketDetailsPage.reducer,
  ticketsPage: fromTicketsPage.reducer,
  tickets: fromTickets.reducer,
  users: fromUsers.reducer,
};

/**
 * Router State
 */
export const selectRouterState = (state: State) => state.router;
export const selectRouterParams = createSelector(
  selectRouterState,
  router => router.state.params,
);

/**
 * Ticket Details Page State
 */
export const selectTicketDetailsPageState = (state: State) => state.ticketDetailsPage;
export const selectIsTicketDetailsPageLoading = createSelector(
  selectTicketDetailsPageState,
  fromTicketDetailsPage.selectIsLoading,
);
export const selectDidTicketDetailsPageFailToLoad = createSelector(
  selectTicketDetailsPageState,
  fromTicketDetailsPage.selectDidFailToLoad,
);

/**
 * Tickets Page State
 */
export const selectTicketsPageState = (state: State) => state.ticketsPage;
export const selectIsTicketsPageLoading = createSelector(
  selectTicketsPageState,
  fromTicketsPage.selectIsLoading,
);
export const selectDidTicketsPageFailToLoad = createSelector(
  selectTicketsPageState,
  fromTicketsPage.selectDidFailToLoad,
);
export const selectTicketDescriptionSearchTerm = createSelector(
  selectTicketsPageState,
  fromTicketsPage.selectDescriptionSearchTerm,
);
export const selectTicketUserIdFilter = createSelector(
  selectTicketsPageState,
  fromTicketsPage.selectUserIdFilter,
);

/**
 * Tickets State
 */
export const selectTicketsState = (state: State) => state.tickets;
export const selectAllTickets = createSelector(
  selectTicketsState,
  fromTickets.selectAll,
);

/**
 * Users State
 */
export const selectUsersState = (state: State) => state.users;
export const selectAllUsers = createSelector(
  selectUsersState,
  fromUsers.selectAll,
);
export const selectUsersEntities = createSelector(
  selectUsersState,
  fromUsers.selectEntities,
);

/**
 * Custom Tickets Selectors
 */
export const selectTicketViews = createSelector(
  selectAllTickets,
  selectUsersEntities,
  (filteredTickets, usersEntities): TicketViewModel[] => {
    return filteredTickets.map(ticket => ({
      ...ticket,
      assignee: usersEntities[ticket.assigneeId],
    }));
  },
);

/**
 * Custom Ticket Details Page Selectors
 */
export const selectActiveTicket = createSelector(
  selectRouterParams,
  selectTicketViews,
  (routerParams, tickets) => {
    return tickets.find(ticket => ticket.id === parseInt(routerParams.ticketId, 10));
  },
);

/**
 * Custom Tickets Page Selectors
 */
export const selectFilteredTickets = createSelector(
  selectTicketViews,
  selectTicketDescriptionSearchTerm,
  selectTicketUserIdFilter,
  (tickets, descriptionSearchTerm, userIdFilter) => {
    return tickets.filter(ticket => {
      const descriptionMatches = ticket.description
        .toLowerCase()
        .includes(descriptionSearchTerm.toLowerCase());

      const userIdMatches = userIdFilter === -1 || ticket.assigneeId === userIdFilter;

      return descriptionMatches && userIdMatches;
    });
  },
);
