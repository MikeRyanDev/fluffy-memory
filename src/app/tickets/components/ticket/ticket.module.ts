import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { MatCardModule, MatCheckboxModule } from '@angular/material';
import { TicketComponent, FirstCharPipe } from './ticket.component';

@NgModule({
  imports: [CommonModule, RouterModule, MatCardModule, MatCheckboxModule],
  declarations: [FirstCharPipe, TicketComponent],
  exports: [TicketComponent],
})
export class TicketComponentModule {}
