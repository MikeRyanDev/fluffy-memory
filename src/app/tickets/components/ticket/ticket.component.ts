import { Component, Input, Output, EventEmitter, Pipe, PipeTransform } from '@angular/core';
import { TicketViewModel } from 'src/app/shared/models';

@Pipe({ name: 'appFirstChar' })
export class FirstCharPipe implements PipeTransform {
  transform(value: string) {
    return value[0].toUpperCase();
  }
}

@Component({
  selector: 'app-ticket',
  template: `
    <mat-card [class.completed]="ticket?.completed">
      <div class="avatar">{{ ticket?.assignee.name | appFirstChar }}</div>
      <a [routerLink]="['/ticket', ticket?.id]">{{ ticket?.description }}</a>

      <mat-checkbox [checked]="ticket?.completed" (change)="onToggleComplete()"></mat-checkbox>
    </mat-card>
  `,
  styles: [
    `
      :host {
        display: block;
      }

      a {
        text-decoration: none;
        color: inherit;
        flex-grow: 1;
        margin-right: 12px;
      }

      mat-card {
        display: flex;
        align-items: center;
      }

      mat-checkbox {
        flex-grow: 0;
        flex-shrink: 0;
      }

      .avatar {
        width: 36px;
        height: 36px;
        line-height: 36px;
        text-align: center;
        vertical-align: middle;
        background-color: #f8bbd0;
        border-radius: 24px;
        margin-right: 12px;
      }

      .completed {
        opacity: 0.7;
      }

      .completed a {
        text-decoration: line-through;
        font-style: italic;
        opacity: 0.7;
      }
    `,
  ],
})
export class TicketComponent {
  @Input() ticket: TicketViewModel | undefined;
  @Output() toggleComplete = new EventEmitter();

  onToggleComplete() {
    this.toggleComplete.emit();
  }
}
