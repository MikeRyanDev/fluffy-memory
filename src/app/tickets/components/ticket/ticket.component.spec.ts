import { NO_ERRORS_SCHEMA } from '@angular/core';
import { TestBed, ComponentFixture } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { RouterTestingModule } from '@angular/router/testing';
import { TicketViewModel } from '../../../shared/models';
import { TicketComponent, FirstCharPipe } from './ticket.component';

describe('Ticket Component', () => {
  const mockTicket: TicketViewModel = {
    id: 3,
    description: 'Another Test Ticket',
    completed: false,
    assigneeId: 111,
    assignee: {
      id: 111,
      name: 'Mock User',
    },
  };

  let fixture: ComponentFixture<TicketComponent>;

  beforeEach(async () => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [TicketComponent, FirstCharPipe],
      schemas: [NO_ERRORS_SCHEMA],
    });

    await TestBed.compileComponents();

    fixture = TestBed.createComponent(TicketComponent);
    fixture.componentInstance.ticket = mockTicket;
    fixture.detectChanges();
  });

  it('should render correctly', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should emit a "toggleComplete" output when the checkbox is toggled', () => {
    const nextCallback = jasmine.createSpy();

    const subscription = fixture.componentInstance.toggleComplete.subscribe(nextCallback);
    fixture.debugElement.query(By.css('mat-checkbox')).triggerEventHandler('change', {});

    expect(nextCallback).toHaveBeenCalled();

    subscription.unsubscribe();
  });

  it('should mark the checkbox as checked if the ticket is complete', () => {
    fixture.componentInstance.ticket = {
      ...mockTicket,
      completed: true,
    };
    fixture.detectChanges();
    const { checked } = fixture.debugElement.query(By.css('mat-checkbox')).properties;

    expect(checked).toBe(true);
  });
});
