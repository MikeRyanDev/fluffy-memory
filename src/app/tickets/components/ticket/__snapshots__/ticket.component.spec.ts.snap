// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticket Component should render correctly 1`] = `
<app-ticket
  ticket={[Function Object]}
  toggleComplete={[Function EventEmitter]}
>
  <mat-card>
    <div
      class="avatar"
    >
      M
    </div>
    <a
      href="/ticket/3"
      ng-reflect-router-link="/ticket,3"
    >
      Another Test Ticket
    </a>
    <mat-checkbox />
  </mat-card>
</app-ticket>
`;
